### PR TITLE
Feature/741 b prisma transactions

### PR DESCRIPTION
**Pull request by Kasshern**
# Description

Implements prototype for prisma service atomic execution of db operations. 

## Related Issue

SRI - Introduce Prisma Transactions #741 

## Motivation and Context

links Db operations atomically 

## How Has This Been Tested

tests forthcoming 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Request to be added as a Code Owner/Maintainer

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] I commit to abide by the Responsibilities of Code Owners/Maintainers.

### CODE REVIEW

File: examples/bri-3/src/bri/workgroup/workflows/agents/workflowsStorage.agent.ts [hunk 1]
```diff
@@ -104,4 +106,24 @@ export class WorkflowStorageAgent {
       where: { id: workflow.id },
     });
   }
+
+  async storeWorkflowTransaction(
+    bpiAccountCandidate: BpiAccount,
+    workflowCandidate: Workflow,
+  ): Promise<any> {
+    const boundStoreNewWorkflow = this.storeNewWorkflow.bind(this);
```
Comment (skosito) [inline examples/bri-3/src/bri/workgroup/workflows/agents/workflowsStorage.agent.ts:114]: why binding this here?

File: examples/bri-3/src/bri/workgroup/workflows/agents/workflowsStorage.agent.ts [hunk 2]
```diff
@@ -104,4 +106,24 @@ export class WorkflowStorageAgent {
...
+    const boundStoreNewBpiAccount =
+      this.bpiAccountStorageAgent.storeNewBpiAccount.bind(
+        this.bpiAccountStorageAgent,
+      );
+    const results = await this.prisma.executeTransaction(() => {
```
Comment (skosito) [inline examples/bri-3/src/bri/workgroup/workflows/agents/workflowsStorage.agent.ts:119]: you are still passing only one operation here, so not quite showing array of sequential operations, it should be array [storeBpiAccount, storeWorkflow], not storeBpiAccount.then(storeWorkflow)? this way it seems we are controlling this execution, and prisma should sequentially execute all or nothing behind the scenes if i understand correctly

File: examples/bri-3/src/bri/workgroup/workflows/agents/workflowsStorage.agent.ts [hunk 3]
```diff
@@ -49,14 +53,14 @@ export class WorkflowStorageAgent {
     });
   }
 
-  async storeNewWorkflow(workflow: Workflow): Promise<Workflow> {
+  storeNewWorkflow(workflow: Workflow): PrismaPromise<WorkflowModel> {
```
Comment (skosito) [inline examples/bri-3/src/bri/workgroup/workflows/agents/workflowsStorage.agent.ts:56]: not sure i understand changes to this method, could you please add more details?

File: examples/bri-3/src/shared/prisma/prisma.service.ts [hunk 4]
```diff
@@ -1,9 +1,23 @@
...
     super({ log: ['info'] });
   }
+
+  public async executeTransaction(
+    ...operations: DbOperation[]
+  ): Promise<any[]> {
+    return await this.$transaction(async () => {
+      const results: any[] = [];
+      for (const operation of operations) {
+        results.push(await operation());
```
Comment (skosito) [inline examples/bri-3/src/shared/prisma/prisma.service.ts:18]: do we need to await operations?

**Comment by ognjenkurtic** [comment]
@Kasshern could we go through this in detail on Monday? would like to get a walk through to be able to better understand some of the decisions made. Or provide async here a little more details please explaining the implementation decisions

**Comment by ognjenkurtic** [comment]
Agreement is to stick with the simple nested query approach for now in few places where we care about transactions until we find a better way to solve this

**Comment by Ybittan** [comment]
Core Dev 02/05/24 - This PR will be replaced with a more simple implementation.

**Pull request closed without merging**